src/cdp/router.spec.ts

```diff
@@ -0,0 +1,104 @@
+import { StubServer } from '../tests/stubServer.spec';
+
+import * as chai from 'chai';
+import chaiAsPromised from 'chai-as-promised';
+chai.use(chaiAsPromised);
+
+import * as sinon from 'sinon';
+
+import { MessageRouter } from './router';
+
+describe('MessageRouter', function () {
+  it('can send a message without a sessionId', async function () {
+    const mockCdpServer = new StubServer();
+    const router = new MessageRouter(mockCdpServer);
+
+    const message = JSON.stringify({ id: 0, method: 'Browser.getVersion' });
+    router.sendMessage(message);
+
+    sinon.assert.calledOnce(mockCdpServer.sendMessage);
+    sinon.assert.calledWith(mockCdpServer.sendMessage, message);
+  });
+
+  it('can send a message with a sessionId', async function () {
+    const mockCdpServer = new StubServer();
+    const router = new MessageRouter(mockCdpServer);
+
+    const message = JSON.stringify({
+      sessionId: 'ABCD',
+      id: 0,
+      method: 'Page.enable',
+    });
+    router.sendMessage(message);
+
+    sinon.assert.calledOnce(mockCdpServer.sendMessage);
+    sinon.assert.calledWith(mockCdpServer.sendMessage, message);
+  });
+
+  it('routes event messages to the correct handler based on sessionId', async function () {
+    const mockCdpServer = new StubServer();
+    const router = new MessageRouter(mockCdpServer);
+
+    const browserMessage = { method: 'Target.attachedToTarget' };
+    const sessionAMessage = { sessionId: 'A', method: 'Page.frameNavigated' };
+    const sessionBMessage = { sessionId: 'B', method: 'Page.loadEventFired' };
+    const onMessage = mockCdpServer.getOnMessage();
+
+    const browserCallback = sinon.fake();
+    const sessionACallback = sinon.fake();
+    const sessionBCallback = sinon.fake();
+
+    // Register for browser message callbacks.
+    router.addClient(null, browserCallback);
+
+    // Verify that the browser callback receives the message.
+    onMessage(JSON.stringify(browserMessage));
+    sinon.assert.calledOnceWithExactly(browserCallback, browserMessage);
+    browserCallback.resetHistory();
```

Comment:
Just a question: Is it intended this step is actually checked in the next verification? It looks like we need only 2 steps:
* `// Send a message for session B and verify that only the session B callback receives it.` 
* `// Send another message for the browser and verify that only the browser callback receives it.`
It looks like you write a test incrementally along with the functionality, and it's perfectly fine. But I wanted to double check that it's intended.

Reply:
Yes, these tests are meant to be incremental. The browser callback is intended to be checked twice. Once during `// Verify that the browser callback receives the message.` and then again during `// Send another message for the browser and verify that only the browser callback receives it.`. The purpose of the second check is to verify that registering a new client doesn't interfere with the existing client. I've added a few comments to clarify.